TestMulticastMessages failing

```
=== RUN TestMulticastMessages
2015/07/24 15:05:43 {"Type":2,"Payload":""}
2015/07/24 15:05:43 {2 }
--- FAIL: TestMulticastMessages (0.01s)
	multicast_test.go:124: Multicast Hello message is incorrectly formatted
=== RUN TestMulticastAnnouncing
--- PASS: TestMulticastAnnouncing (1.10s)
FAIL
coverage: 64.2% of statements
exit status 1
FAIL	github.com/getlantern/multicast	3.335s
make: *** [test-and-cover] Error 1
```

I think this was mentioned already in https://github.com/getlantern/lantern/pull/2646, but I'm consistently seeing this test fail.

--- a/src/github.com/getlantern/multicast/multicast_test.go
+++ b/src/github.com/getlantern/multicast/multicast_test.go
@@ -112,19 +112,30 @@ func TestMulticastMessages(t *testing.T) {
 	}
 
 	b := make([]byte, messageMaxSize)
-	n, _, e := mc2.read(b)
-	if e != nil {
-		t.Fatal("Unable to multicast message")
-	}
-	if n > 0 {
-		var msg multicastMessage
-		if e = json.Unmarshal(b[:n], &msg); e != nil || msg.Type != typeHello || msg.Payload != "testHello" {
-			stdLog.Println(string(b[:n]))
-			stdLog.Println(msg)
-			t.Fatal("Multicast Hello message is incorrectly formatted")
+	var msg multicastMessage
+Out:
+	for {
+		if n, _, err := mc2.read(b); err != nil {
+			t.Fatal("Error reading multicast message")
+		} else {
+			if err := json.Unmarshal(b[:n], &msg); err != nil {
+				t.Fatal("Error unmarshalling multicast message")
+			} else {
+				switch msg.Type {
+				case typeHello:
+					if msg.Payload != "testHello" {
+						stdLog.Println(string(b[:n]))
+						stdLog.Println(msg)
+						t.Fatal("Multicast Hello message is incorrectly formatted")
+					}
+					break Out
+				case typeBye:
+					continue
+				default:
+					t.Fatal("Unknown multicast message type")
+				}
+			}
 		}
-	} else {
-		stdLog.Println("Received 0 bytes")
 	}
 
 	mc1.LeaveMulticast()

